File: android/app/src/main/java/com/studyapp/sync/SyncDeltaSerializer.kt
package com.studyapp.sync

import com.studyapp.domain.model.Exam
import com.studyapp.domain.model.Goal
import com.studyapp.domain.model.Material
import com.studyapp.domain.model.PlanItem
import com.studyapp.domain.model.ProblemReviewRecord
import com.studyapp.domain.model.StudyPlan
import com.studyapp.domain.model.StudySession
import com.studyapp.domain.model.Subject
import com.studyapp.domain.model.TimetableEntry
import com.studyapp.domain.model.TimetablePeriod
import com.studyapp.domain.model.TimetableReviewRecord
import com.studyapp.domain.model.TimetableTerm
import com.studyapp.domain.usecase.AppData
import com.studyapp.domain.usecase.PlanData
import com.studyapp.domain.usecase.toExam
import com.studyapp.domain.usecase.toGoal
import com.studyapp.domain.usecase.toJson
import com.studyapp.domain.usecase.toMaterial
import com.studyapp.domain.usecase.toPlanItem
import com.studyapp.domain.usecase.toProblemReviewRecord
import com.studyapp.domain.usecase.toStudyPlan
import com.studyapp.domain.usecase.toStudySession
import com.studyapp.domain.usecase.toSubject
import com.studyapp.domain.usecase.toTimetableEntry
import com.studyapp.domain.usecase.toTimetablePeriod
import com.studyapp.domain.usecase.toTimetableReviewRecord
import com.studyapp.domain.usecase.toTimetableTerm
import org.json.JSONObject

object SyncDeltaSerializer {
    fun decompose(appData: AppData): List<SyncEntityEnvelope> {
        val envelopes = ArrayList<SyncEntityEnvelope>(estimatedEnvelopeCount(appData))
        appData.subjects.forEach { envelopes += envelope(it, SyncEntityKind.SUBJECT) }
        appData.materials.forEach { envelopes += envelope(it, SyncEntityKind.MATERIAL) }
        appData.sessions.forEach { envelopes += envelope(it, SyncEntityKind.SESSION) }
        appData.goals.forEach { envelopes += envelope(it, SyncEntityKind.GOAL) }
        appData.exams.forEach { envelopes += envelope(it, SyncEntityKind.EXAM) }
        appData.plans.forEach { planData ->
            envelopes += envelope(planData.plan, SyncEntityKind.PLAN)
            planData.items.forEach { envelopes += envelope(it, SyncEntityKind.PLAN_ITEM) }
        }
        appData.timetablePeriods.forEach { envelopes += envelope(it, SyncEntityKind.TIMETABLE_PERIOD) }
        appData.timetableEntries.forEach { envelopes += envelope(it, SyncEntityKind.TIMETABLE_ENTRY) }
        appData.timetableTerms.forEach { envelopes += envelope(it, SyncEntityKind.TIMETABLE_TERM) }
        appData.timetableReviewRecords.forEach { envelopes += envelope(it, SyncEntityKind.TIMETABLE_REVIEW_RECORD) }
        appData.problemReviewRecords.forEach { envelopes += envelope(it, SyncEntityKind.PROBLEM_REVIEW_RECORD) }
        return envelopes
    }

    fun changedSince(appData: AppData, cursor: Long): List<SyncEntityEnvelope> {
        return decompose(appData).filter { it.updatedAt > cursor }
    }

    fun assemble(envelopes: List<SyncEntityEnvelope>, onto: AppData): AppData {
        val partial = partialAppData(envelopes, onto.exportDate)
        return SyncMergeEngine.merge(local = onto, remote = partial)
    }

    private fun estimatedEnvelopeCount(appData: AppData): Int {
        return appData.subjects.size +
            appData.materials.size +
            appData.sessions.size +
            appData.goals.size +
            appData.exams.size +
            appData.plans.size +
            appData.plans.sumOf { it.items.size } +
            appData.timetablePeriods.size +
            appData.timetableEntries.size +
            appData.timetableTerms.size +
            appData.timetableReviewRecords.size +
            appData.problemReviewRecords.size
    }

    private fun envelope(
        syncId: String,
        updatedAt: Long,
        deletedAt: Long?,
        kind: SyncEntityKind,
        json: String
    ): SyncEntityEnvelope {
        return SyncEntityEnvelope(kind = kind, syncId = syncId, updatedAt = updatedAt, deletedAt = deletedAt, json = json)
    }

    private fun envelope(subject: Subject, kind: SyncEntityKind) =
        envelope(subject.syncId, subject.updatedAt, subject.deletedAt, kind, subject.toJson().toString())

    private fun envelope(material: Material, kind: SyncEntityKind) =
        envelope(material.syncId, material.updatedAt, material.deletedAt, kind, material.toJson().toString())

    private fun envelope(session: StudySession, kind: SyncEntityKind) =
        envelope(session.syncId, session.updatedAt, session.deletedAt, kind, session.toJson().toString())

    private fun envelope(goal: Goal, kind: SyncEntityKind) =
        envelope(goal.syncId, goal.updatedAt, goal.deletedAt, kind, goal.toJson().toString())

    private fun envelope(exam: Exam, kind: SyncEntityKind) =
        envelope(exam.syncId, exam.updatedAt, exam.deletedAt, kind, exam.toJson().toString())

    private fun envelope(plan: StudyPlan, kind: SyncEntityKind) =
        envelope(plan.syncId, plan.updatedAt, plan.deletedAt, kind, plan.toJson().toString())

    private fun envelope(item: PlanItem, kind: SyncEntityKind) =
        envelope(item.syncId, item.updatedAt, item.deletedAt, kind, item.toJson().toString())

    private fun envelope(period: TimetablePeriod, kind: SyncEntityKind) =
        envelope(period.syncId, period.updatedAt, period.deletedAt, kind, period.toJson().toString())

    private fun envelope(entry: TimetableEntry, kind: SyncEntityKind) =
        envelope(entry.syncId, entry.updatedAt, entry.deletedAt, kind, entry.toJson().toString())

    private fun envelope(term: TimetableTerm, kind: SyncEntityKind) =
        envelope(term.syncId, term.updatedAt, term.deletedAt, kind, term.toJson().toString())

    private fun envelope(record: TimetableReviewRecord, kind: SyncEntityKind) =
        envelope(record.syncId, record.updatedAt, record.deletedAt, kind, record.toJson().toString())

    private fun envelope(record: ProblemReviewRecord, kind: SyncEntityKind) =
        envelope(record.syncId, record.updatedAt, record.deletedAt, kind, record.toJson().toString())

    private fun partialAppData(envelopes: List<SyncEntityEnvelope>, exportDate: Long): AppData {
        val subjects = mutableListOf<Subject>()
        val materials = mutableListOf<Material>()
        val sessions = mutableListOf<StudySession>()
        val goals = mutableListOf<Goal>()
        val exams = mutableListOf<Exam>()
        val plans = mutableListOf<StudyPlan>()
        val planItems = mutableListOf<PlanItem>()
        val timetablePeriods = mutableListOf<TimetablePeriod>()
        val timetableEntries = mutableListOf<TimetableEntry>()
        val timetableTerms = mutableListOf<TimetableTerm>()
        val timetableReviewRecords = mutableListOf<TimetableReviewRecord>()
        val problemReviewRecords = mutableListOf<ProblemReviewRecord>()

        envelopes.forEach { envelope ->
            runCatching {
                val json = JSONObject(envelope.json)
                when (envelope.kind) {
                    SyncEntityKind.SUBJECT -> subjects += json.toSubject()
                    SyncEntityKind.MATERIAL -> materials += json.toMaterial()
                    SyncEntityKind.SESSION -> sessions += json.toStudySession()
                    SyncEntityKind.GOAL -> goals += json.toGoal()
                    SyncEntityKind.EXAM -> exams += json.toExam()
                    SyncEntityKind.PLAN -> plans += json.toStudyPlan()
                    SyncEntityKind.PLAN_ITEM -> planItems += json.toPlanItem()
                    SyncEntityKind.TIMETABLE_PERIOD -> timetablePeriods += json.toTimetablePeriod()
                    SyncEntityKind.TIMETABLE_ENTRY -> timetableEntries += json.toTimetableEntry()
                    SyncEntityKind.TIMETABLE_TERM -> timetableTerms += json.toTimetableTerm()
                    SyncEntityKind.TIMETABLE_REVIEW_RECORD -> timetableReviewRecords += json.toTimetableReviewRecord()
                    SyncEntityKind.PROBLEM_REVIEW_RECORD -> json.toProblemReviewRecord()?.let { problemReviewRecords += it }
                }
            }
        }

        val itemsByPlanSyncId = planItems.groupBy { it.planSyncId }
        val planData = plans.map { plan -> PlanData(plan = plan, items = itemsByPlanSyncId[plan.syncId].orEmpty()) }

        return AppData(
            subjects = subjects,
            materials = materials,
            sessions = sessions,
            goals = goals,
            exams = exams,
            plans = planData,
            timetablePeriods = timetablePeriods,
            timetableEntries = timetableEntries,
            timetableTerms = timetableTerms,
            timetableReviewRecords = timetableReviewRecords,
            problemReviewRecords = problemReviewRecords,
            exportDate = exportDate
        )
    }
}

File: ios/StudyApp/Data/Sync/SyncDeltaSerializer.swift
import Foundation

/// Identifies which domain entity an envelope carries. Stable rawValues are
/// used as part of the Firestore document ID, so renaming or removing cases
/// is a breaking change.
enum SyncEntityKind: String, Codable, CaseIterable {
    case subject
    case material
    case session
    case goal
    case exam
    case plan
    case planItem
    case timetablePeriod
    case timetableEntry
    case timetableTerm
    case timetableReviewRecord
    case problemReviewRecord
}

/// A single entity, encoded and paired with the metadata we need for
/// incremental sync (conflict resolution, tombstone propagation, delta cursor).
///
/// The envelope carries the entity as JSON rather than as a `Codable`
/// existential so the storage backend (Firestore, tests, local caches) can
/// persist it verbatim without needing the domain types.
struct SyncEntityEnvelope: Equatable {
    var kind: SyncEntityKind
    /// The entity's stable `syncId` (matches `Subject.syncId`, `Material.syncId`, etc.).
    var syncId: String
    /// Client-side last-modified timestamp. Authoritative for merge ordering.
    var updatedAt: Int64
    /// Tombstone timestamp when the entity is logically deleted.
    var deletedAt: Int64?
    /// The entity encoded as JSON using the domain type's `Codable` conformance.
    var json: String
    /// Optional revision metadata (ignored by legacy clients/readers).
    var revisionId: String?
    var parentRevisionId: String?
    var deviceId: String?
    var contentHash: String?

    /// Stable document id used by the Firestore delta store. The leading kind
    /// keeps ids unique across domain types even if syncIds collide (they
    /// shouldn't, but belt-and-braces).
    var documentId: String {
        "\(kind.rawValue)-\(syncId)"
    }
}

/// Pure serializer / reassembler between `AppData` and a bag of
/// `SyncEntityEnvelope`s. Swapping the Firestore layer for any other
/// delta-capable store is a matter of writing/reading envelopes.
///
/// Design notes:
/// * `decompose` emits one envelope per entity (including tombstones). Plans
///   and their items are emitted as separate kinds so each has its own
///   `updatedAt` and can sync independently.
/// * `assemble` rebuilds a full `AppData` by merging the envelopes *onto*
///   a base snapshot using `SyncMergeEngine`, so last-writer-wins with
///   tombstone precedence is preserved.
/// * `changedSince` is the primary upload primitive: hand it the local
///   `AppData` and the cursor from the previous successful sync, and it
///   returns exactly the envelopes that need to be pushed.
enum SyncDeltaSerializer {

    // MARK: - Decomposition (AppData -> envelopes)

    static func decompose(_ appData: AppData) -> [SyncEntityEnvelope] {
        var envelopes: [SyncEntityEnvelope] = []
        envelopes.reserveCapacity(estimatedEnvelopeCount(for: appData))

        for value in appData.subjects {
            envelopes.append(envelope(for: value, kind: .subject))
        }
        for value in appData.materials {
            envelopes.append(envelope(for: value, kind: .material))
        }
        for value in appData.sessions {
            envelopes.append(envelope(for: value, kind: .session))
        }
        for value in appData.goals {
            envelopes.append(envelope(for: value, kind: .goal))
        }
        for value in appData.exams {
            envelopes.append(envelope(for: value, kind: .exam))
        }
        for planData in appData.plans {
            envelopes.append(envelope(for: planData.plan, kind: .plan))
            for item in planData.items {
                envelopes.append(envelope(for: item, kind: .planItem))
            }
        }
        for value in appData.timetablePeriods {
            envelopes.append(envelope(for: value, kind: .timetablePeriod))
        }
        for value in appData.timetableEntries {
            envelopes.append(envelope(for: value, kind: .timetableEntry))
        }
        for value in appData.timetableTerms {
            envelopes.append(envelope(for: value, kind: .timetableTerm))
        }
        for value in appData.timetableReviewRecords {
            envelopes.append(envelope(for: value, kind: .timetableReviewRecord))
        }
        for value in appData.problemReviewRecords {
            envelopes.append(envelope(for: value, kind: .problemReviewRecord))
        }
        return envelopes
    }

    /// Same as `decompose`, but filtered to only entities after the composite cursor.
    static func changedSince(_ appData: AppData, cursor: SyncDeltaCursor) -> [SyncEntityEnvelope] {
        decompose(appData).filter { $0.cursorPosition > cursor }
    }

    /// Legacy Int64 cursor support.
    static func changedSince(_ appData: AppData, cursor: Int64) -> [SyncEntityEnvelope] {
        changedSince(appData, cursor: SyncDeltaCursor.fromLegacy(cursor))
    }

    // MARK: - Reassembly (envelopes + base -> AppData)

    /// Merges the given envelopes onto `base` using `SyncMergeEngine`
    /// semantics (last-writer-wins with tombstone precedence and
    /// problem-progress preservation).
    ///
    /// Envelopes that fail to decode are skipped; we log via the caller
    /// rather than `fatalError` so a single bad document does not prevent
    /// the rest of the sync from applying.
    static func assemble(envelopes: [SyncEntityEnvelope], onto base: AppData) -> AppData {
        let partial = partialAppData(from: envelopes, exportDate: base.exportDate)
        return SyncMergeEngine.merge(local: base, remote: partial)
    }

    // MARK: - Private

    private static func estimatedEnvelopeCount(for appData: AppData) -> Int {
        appData.subjects.count
            + appData.materials.count
            + appData.sessions.count
            + appData.goals.count
            + appData.exams.count
            + appData.plans.count
            + appData.plans.reduce(0) { $0 + $1.items.count }
            + appData.timetablePeriods.count
            + appData.timetableEntries.count
            + appData.timetableTerms.count
            + appData.timetableReviewRecords.count
            + appData.problemReviewRecords.count
    }

    /// Packs a single `Encodable` domain entity (including its `updatedAt`
    /// and `deletedAt`) into an envelope. JSON encoding is expected to
    /// succeed for the project's own Codable types; a failure returns a
    /// defensively-empty envelope rather than throwing, which keeps
    /// decompose infallible.
    private static func envelope<T: SyncDeltaEntity>(for entity: T, kind: SyncEntityKind) -> SyncEntityEnvelope {
        let json = (try? encodeToString(entity)) ?? "{}"
        return SyncEntityEnvelope(
            kind: kind,
            syncId: entity.syncId,
            updatedAt: entity.updatedAt,
            deletedAt: entity.deletedAt,
            json: json
        )
    }

    private static func encodeToString<T: Encodable>(_ value: T) throws -> String {
        let encoder = JSONEncoder()
        let data = try encoder.encode(value)
        return String(data: data, encoding: .utf8) ?? "{}"
    }

    private static func decodeFromString<T: Decodable>(_ string: String, as: T.Type) throws -> T {
        let data = Data(string.utf8)
        return try JSONDecoder().decode(T.self, from: data)
    }

    /// Rebuilds the subset of `AppData` that the envelopes describe.
    /// The resulting `AppData` is suitable as the `remote` argument to
    /// `SyncMergeEngine.merge`. Unknown kinds are skipped.
    static func partialAppData(
        from envelopes: [SyncEntityEnvelope],
        exportDate: Int64
    ) -> AppData {
        var subjects: [Subject] = []
        var materials: [Material] = []
        var sessions: [StudySession] = []
        var goals: [Goal] = []
        var exams: [Exam] = []
        var plans: [StudyPlan] = []
        var planItems: [PlanItem] = []
        var timetablePeriods: [TimetablePeriod] = []
        var timetableEntries: [TimetableEntry] = []
        var timetableTerms: [TimetableTerm] = []
        var timetableReviewRecords: [TimetableReviewRecord] = []
        var problemReviewRecords: [ProblemReviewRecord] = []

        for envelope in envelopes {
            switch envelope.kind {
            case .subject:
                if let value = try? decodeFromString(envelope.json, as: Subject.self) {
                    subjects.append(value)
                }
            case .material:
                if let value = try? decodeFromString(envelope.json, as: Material.self) {
                    materials.append(value)
                }
            case .session:
                if let value = try? decodeFromString(envelope.json, as: StudySession.self) {
                    sessions.append(value)
                }
            case .goal:
                if let value = try? decodeFromString(envelope.json, as: Goal.self) {
                    goals.append(value)
                }
            case .exam:
                if let value = try? decodeFromString(envelope.json, as: Exam.self) {
                    exams.append(value)
                }
            case .plan:
                if let value = try? decodeFromString(envelope.json, as: StudyPlan.self) {
                    plans.append(value)
                }
            case .planItem:
                if let value = try? decodeFromString(envelope.json, as: PlanItem.self) {
                    planItems.append(value)
                }
            case .timetablePeriod:
                if let value = try? decodeFromString(envelope.json, as: TimetablePeriod.self) {
                    timetablePeriods.append(value)
                }
            case .timetableEntry:
                if let value = try? decodeFromString(envelope.json, as: TimetableEntry.self) {
                    timetableEntries.append(value)
                }
            case .timetableTerm:
                if let value = try? decodeFromString(envelope.json, as: TimetableTerm.self) {
                    timetableTerms.append(value)
                }
            case .timetableReviewRecord:
                if let value = try? decodeFromString(envelope.json, as: TimetableReviewRecord.self) {
                    timetableReviewRecords.append(value)
                }
            case .problemReviewRecord:
                if let value = try? decodeFromString(envelope.json, as: ProblemReviewRecord.self) {
                    problemReviewRecords.append(value)
                }
            }
        }

        // Rebuild plans as [PlanData] by grouping plan items under their plan's syncId.
        let itemsByPlanSyncId = Dictionary(grouping: planItems, by: \.planSyncId)
        let planData: [PlanData] = plans.map { plan in
            let items = (itemsByPlanSyncId[plan.syncId] ?? [])
            return PlanData(plan: plan, items: items)
        }

        return AppData(
            subjects: subjects,
            materials: materials,
            sessions: sessions,
            goals: goals,
            exams: exams,
            plans: planData,
            timetablePeriods: timetablePeriods,
            timetableEntries: timetableEntries,
            timetableTerms: timetableTerms,
            timetableReviewRecords: timetableReviewRecords,
            problemReviewRecords: problemReviewRecords,
            exportDate: exportDate
        )
    }
}

/// Internal conformance used by `SyncDeltaSerializer.envelope(for:kind:)` so
/// we can encode any of the domain entities through a single code path. All
/// participating types already declare `syncId`, `updatedAt`, and `deletedAt`;
/// this protocol just gives the compiler a uniform handle.
protocol SyncDeltaEntity: Encodable {
    var syncId: String { get }
    var updatedAt: Int64 { get }
    var deletedAt: Int64? { get }
}

extension Subject: SyncDeltaEntity {}
extension Material: SyncDeltaEntity {}
extension StudySession: SyncDeltaEntity {}
extension Goal: SyncDeltaEntity {}
extension Exam: SyncDeltaEntity {}
extension StudyPlan: SyncDeltaEntity {}
extension PlanItem: SyncDeltaEntity {}
extension TimetablePeriod: SyncDeltaEntity {}
extension TimetableEntry: SyncDeltaEntity {}
extension TimetableTerm: SyncDeltaEntity {}
extension TimetableReviewRecord: SyncDeltaEntity {}
extension ProblemReviewRecord: SyncDeltaEntity {}
